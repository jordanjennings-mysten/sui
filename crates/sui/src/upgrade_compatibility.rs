#![allow(dead_code)]
#![allow(unused_imports)]
#![allow(unused_variables)]

use std::collections::BTreeSet;
use move_binary_format::compatibility::{Compatibility};
use move_binary_format::compatibility_mode::{CompatibilityMode};
use move_binary_format::normalized::{Enum, Function, Struct};
use move_core_types::account_address::AccountAddress;
use move_core_types::identifier::{IdentStr, Identifier};
use thiserror::Error;
use anyhow::{anyhow};
use move_core_types::language_storage::ModuleId;
use sui_indexer::config::NameServiceOptions;

/// Errors that can occur during upgrade compatibility checks.
#[derive(Debug, Error)]
enum UpgradeCompatibilityModeError {
    #[error("Struct missing: {}", name.as_str())]
    StructMissing {
        name: Identifier,
        old_struct: Struct,
    },
    #[error("Struct ability mismatch: {}\n  Old: {:?}\n  New: {:?}", name.as_str(), old_struct.abilities, new_struct.abilities)]
    StructAbilityMismatch {
        name: Identifier,
        old_struct: Struct,
        new_struct: Struct,
    },
    #[error("Struct type param mismatch: {}\n  Old: {:?}\n  New: {:?}", name.as_str(), old_struct.type_parameters, new_struct.type_parameters)]
    StructTypeParamMismatch {
        name: Identifier,
        old_struct: Struct,
        new_struct: Struct,
    },
    #[error("Struct field mismatch: {}\n  Old: {:?}\n  New: {:?}", name.as_str(), old_struct.fields, new_struct.fields)]
    StructFieldMismatch {
        name: Identifier,
        old_struct: Struct,
        new_struct: Struct,
    },
    #[error("Enum missing: {}\n  {:?}", name.as_str(), old_enum)]
    EnumMissing {
        name: Identifier,
        old_enum: Enum,
    },
    #[error("Enum ability mismatch: {}\n  Old: {:?}\n  New: {:?}", name.as_str(), old_enum.abilities, new_enum.abilities)]
    EnumAbilityMismatch {
        name: Identifier,
        old_enum: Enum,
        new_enum: Enum,
    },
    #[error("Enum type param mismatch: {}\n  Old: {:?}\n  New: {:?}", name.as_str(), old_enum.type_parameters, new_enum.type_parameters)]
    EnumTypeParamMismatch {
        name: Identifier,
        old_enum: Enum,
        new_enum: Enum,
    },
    #[error("Enum new variant: {}\n  Old: {:?}\n  New: {:?}", name.as_str(), old_enum.variants, new_enum.variants)]
    EnumNewVariant {
        name: Identifier,
        old_enum: Enum,
        new_enum: Enum,
    },
    #[error("Enum variant missing: {}\n  {:?}", name.as_str(), old_enum.variants[*tag])]
    EnumVariantMissing {
        name: Identifier,
        old_enum: Enum,
        tag: usize,
    },
    #[error("Enum variant mismatch: {}\n  Old: {:?}\n  New: {:?}", name.as_str(), old_enum.variants[*tag], new_enum.variants[*tag])]
    EnumVariantMismatch {
        name: Identifier,
        old_enum: Enum,
        new_enum: Enum,
        tag: usize,
    },
    #[error("Function missing public: {}", name.as_str())]
    FunctionMissingPublic {
        name: Identifier,
        old_function: Function,
    },
    #[error("Function missing friend: {}", name.as_str())]
    FunctionMissingFriend {
        name: Identifier,
        old_function: Function,
    },
    #[error("Function missing entry: {}", name.as_str())]
    FunctionMissingEntry {
        name: Identifier,
        old_function: Function,
    },
    #[error("Function signature mismatch: {}\n  Old:\n    Params: {:?}\n    Return: {:?}\n  New:\n    Params: {:?}\n    Return: {:?}", name.as_str(),
        old_function.parameters, old_function.return_, new_function.parameters, new_function.return_
    )]
    FunctionSignatureMismatch {
        name: Identifier,
        old_function: Function,
        new_function: Function,
    },
    #[error("Function lost public visibility: {}", name.as_str())]
    FunctionLostPublicVisibility {
        name: Identifier,
        old_function: Function,
    },
    #[error("Function lost friend visibility: {}", name.as_str())]
    FunctionLostFriendVisibility {
        name: Identifier,
        old_function: Function,
    },
    #[error("Function entry compatibility: {}\n  Old: {:?}\n  New: {:?}", name.as_str(), old_function, new_function)]
    FunctionEntryCompatibility {
        name: Identifier,
        old_function: Function,
        new_function: Function,
    },
    #[error("Friend module missing: {0:?} {1:?}")]
    FriendModuleMissing(BTreeSet<ModuleId>, BTreeSet<ModuleId>),
}

/// A compatibility mode that collects errors
pub struct CliCompatibilityMode {
    errors: Vec<UpgradeCompatibilityModeError>,
}

impl Default for CliCompatibilityMode {
    fn default() -> Self {
        Self {
            errors: vec![],
        }
    }
}

impl CompatibilityMode for CliCompatibilityMode {
    type Error = anyhow::Error;
    // ignored, address is not populated pre-tx
    fn module_id_mismatch(&mut self, _old_addr: &AccountAddress, _old_name: &IdentStr, _new_addr: &AccountAddress, _new_name: &IdentStr) {}

    fn struct_missing(&mut self, name: &Identifier, old_struct: &Struct) {
        println!("WOT");
        self.errors.push(UpgradeCompatibilityModeError::StructMissing{
            name: name.clone(),
            old_struct: old_struct.clone()
        });
    }

    fn struct_ability_mismatch(&mut self, name: &Identifier, old_struct: &Struct, new_struct: &Struct) {
        self.errors.push(UpgradeCompatibilityModeError::StructAbilityMismatch{
            name: name.clone(),
            old_struct: old_struct.clone(),
            new_struct: new_struct.clone()
        });
    }

    fn struct_type_param_mismatch(&mut self, name: &Identifier, old_struct: &Struct, new_struct: &Struct) {
        self.errors.push(UpgradeCompatibilityModeError::StructTypeParamMismatch {
            name: name.clone(),
            old_struct: old_struct.clone(),
            new_struct: new_struct.clone()
        });
    }

    fn struct_field_mismatch(&mut self, name: &Identifier, old_struct: &Struct, new_struct: &Struct) {
        self.errors.push(UpgradeCompatibilityModeError::StructFieldMismatch {
            name: name.clone(),
            old_struct: old_struct.clone(),
            new_struct: new_struct.clone()
        });
    }

    fn enum_missing(&mut self, name: &Identifier, old_enum: &Enum) {
        self.errors.push(UpgradeCompatibilityModeError::EnumMissing {
            name: name.clone(),
            old_enum: old_enum.clone()
        });
    }

    fn enum_ability_mismatch(&mut self, name: &Identifier, old_enum: &Enum, new_enum: &Enum) {
        self.errors.push(UpgradeCompatibilityModeError::EnumAbilityMismatch {
            name: name.clone(),
            old_enum: old_enum.clone(),
            new_enum: new_enum.clone()
        });
    }

    fn enum_type_param_mismatch(&mut self, name: &Identifier, old_enum: &Enum, new_enum: &Enum) {
        self.errors.push(UpgradeCompatibilityModeError::EnumTypeParamMismatch {
            name: name.clone(),
            old_enum: old_enum.clone(),
            new_enum: new_enum.clone()
        });
    }

    fn enum_new_variant(&mut self, name: &Identifier, old_enum: &Enum, new_enum: &Enum) {
        self.errors.push(UpgradeCompatibilityModeError::EnumNewVariant {
            name: name.clone(),
            old_enum: old_enum.clone(),
            new_enum: new_enum.clone()
        });
    }

    fn enum_variant_missing(&mut self, name: &Identifier, old_enum: &Enum, tag: usize) {
        self.errors.push(UpgradeCompatibilityModeError::EnumVariantMissing {
            name: name.clone(),
            old_enum: old_enum.clone(),
            tag,
        });
    }

    fn enum_variant_mismatch(&mut self, name: &Identifier, old_enum: &Enum, new_enum: &Enum, variant_idx: usize) {
        self.errors.push(UpgradeCompatibilityModeError::EnumVariantMismatch {
            name: name.clone(),
            old_enum: old_enum.clone(),
            new_enum: new_enum.clone(),
            tag: variant_idx
        });
    }

    fn function_missing_public(&mut self, name: &Identifier, old_function: &Function) {
        self.errors.push(UpgradeCompatibilityModeError::FunctionMissingPublic {
            name: name.clone(),
            old_function: old_function.clone()
        });
    }

    fn function_missing_friend(&mut self, name: &Identifier, old_function: &Function) {
        self.errors.push(UpgradeCompatibilityModeError::FunctionMissingFriend {
            name: name.clone(),
            old_function: old_function.clone()
        });
    }

    fn function_missing_entry(&mut self, name: &Identifier, old_function: &Function) {
        self.errors.push(UpgradeCompatibilityModeError::FunctionMissingEntry {
            name: name.clone(),
            old_function: old_function.clone()
        });
    }

    fn function_signature_mismatch(&mut self, name: &Identifier, old_function: &Function, new_function: &Function) {
        self.errors.push(UpgradeCompatibilityModeError::FunctionSignatureMismatch {
            name: name.clone(),
            old_function: old_function.clone(),
            new_function: new_function.clone()
        });
    }

    fn function_lost_public_visibility(&mut self, name: &Identifier, old_function: &Function) {
        self.errors.push(UpgradeCompatibilityModeError::FunctionLostPublicVisibility {
            name: name.clone(),
            old_function: old_function.clone()
        });
    }

    fn function_lost_friend_visibility(&mut self, name: &Identifier, old_function: &Function) {
        self.errors.push(UpgradeCompatibilityModeError::FunctionLostFriendVisibility {
            name: name.clone(),
            old_function: old_function.clone()
        });
    }

    fn function_entry_compatibility(&mut self, name: &Identifier, old_function: &Function, new_function: &Function) {
        self.errors.push(UpgradeCompatibilityModeError::FunctionEntryCompatibility {
            name: name.clone(),
            old_function: old_function.clone(),
            new_function: new_function.clone()
        });
    }

    fn friend_module_missing(&mut self, old_modules: BTreeSet<ModuleId>, new_modules: BTreeSet<ModuleId>) {
        self.errors.push(UpgradeCompatibilityModeError::FriendModuleMissing(old_modules.clone(), new_modules.clone()));
    }

    fn finish(&self, _: &Compatibility) -> Result<(), Self::Error> {
        if !self.errors.is_empty() {
            let errors: Vec<String> = self.errors.iter().map(|e| format!("- {}", e)).collect();
            return Err(anyhow!("Upgrade compatibility check failed with the following errors:\n{}", errors.join("\n")));
        }
        Ok(())
    }
}