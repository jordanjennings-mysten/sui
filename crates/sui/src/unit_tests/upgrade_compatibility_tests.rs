// Copyright (c) Mysten Labs, Inc.
// SPDX-License-Identifier: Apache-2.0

use crate::upgrade_compatibility::compare_packages;
use move_binary_format::CompiledModule;
use std::path::PathBuf;
use sui_move_build::BuildConfig;

#[test]
fn test_all_fail() {
    let (pkg_v1, pkg_v2) = get_packages("all");

    let result = compare_packages(pkg_v1, pkg_v2);
    assert!(result.is_err());
    let err = result.unwrap_err();

    assert_eq!(
        err.to_string(),
        r#"Upgrade compatibility check failed with the following errors:
- Struct ability mismatch: StructAbilityMismatchAdd
  Old: []
  New: [Copy, ]
- Struct ability mismatch: StructAbilityMismatchChange
  Old: [Copy, ]
  New: [Drop, ]
- Struct ability mismatch: StructAbilityMismatchRemove
  Old: [Copy, ]
  New: []
- Struct field mismatch: StructFieldMismatchAdd
  Old: [Field { name: Identifier("a"), type_: U64 }, Field { name: Identifier("b"), type_: U64 }]
  New: [Field { name: Identifier("a"), type_: U64 }, Field { name: Identifier("b"), type_: U64 }, Field { name: Identifier("c"), type_: U64 }]
- Struct field mismatch: StructFieldMismatchChange
  Old: [Field { name: Identifier("a"), type_: U64 }, Field { name: Identifier("b"), type_: U64 }]
  New: [Field { name: Identifier("a"), type_: U64 }, Field { name: Identifier("b"), type_: U8 }]
- Struct field mismatch: StructFieldMismatchRemove
  Old: [Field { name: Identifier("a"), type_: U64 }, Field { name: Identifier("b"), type_: U64 }]
  New: [Field { name: Identifier("a"), type_: U64 }]
- Struct missing: StructToBeRemoved
- Struct type param mismatch: StructTypeParamMismatch
  Old: [DatatypeTyParameter { constraints: [], is_phantom: false }, DatatypeTyParameter { constraints: [], is_phantom: false }]
  New: [DatatypeTyParameter { constraints: [], is_phantom: false }]
- Enum ability mismatch: EnumAbilityMismatchAdd
  Old: []
  New: [Copy, ]
- Enum ability mismatch: EnumAbilityMismatchChange
  Old: [Copy, ]
  New: [Drop, ]
- Enum ability mismatch: EnumAbilityMismatchRemove
  Old: [Copy, ]
  New: []
- Enum new variant: EnumNewVariant
  Old: [Variant { name: Identifier("A"), fields: [] }, Variant { name: Identifier("B"), fields: [] }, Variant { name: Identifier("C"), fields: [] }]
  New: [Variant { name: Identifier("A"), fields: [] }, Variant { name: Identifier("B"), fields: [] }, Variant { name: Identifier("C"), fields: [] }, Variant { name: Identifier("D"), fields: [] }]
- Enum missing: EnumToBeRemoved
  Enum { abilities: [], type_parameters: [], variants: [Variant { name: Identifier("A"), fields: [] }, Variant { name: Identifier("B"), fields: [] }] }
- Enum variant missing: EnumVariantMissing
  Variant { name: Identifier("B"), fields: [] }
- Function signature mismatch: function_add_arg
  Old:
    Params: []
    Return: []
  New:
    Params: [U64]
    Return: []
- Function signature mismatch: function_change_arg
  Old:
    Params: [U64]
    Return: []
  New:
    Params: [U8]
    Return: []
- Function signature mismatch: function_remove_arg
  Old:
    Params: [U64]
    Return: []
  New:
    Params: []
    Return: []
- Function lost public visibility: function_to_have_public_removed"#
    )
}

#[test]
fn test_struct_missing() {
    let (pkg_v1, pkg_v2) = get_packages("struct_missing");
    let result = compare_packages(pkg_v1, pkg_v2);

    assert!(result.is_err());
    let err = result.unwrap_err();
    assert_eq!(err.to_string(), "Upgrade compatibility check failed with the following errors:\n- Struct missing: StructToBeRemoved");
}

#[test]
fn test_friend_link_ok() {
    let (pkg_v1, pkg_v2) = get_packages("friend_linking");
    // upgrade compatibility ignores friend linking
    assert!(compare_packages(pkg_v1, pkg_v2).is_ok());
}

#[test]
fn test_entry_linking_ok() {
    let (pkg_v1, pkg_v2) = get_packages("entry_linking");
    // upgrade compatibility ignores entry linking
    assert!(compare_packages(pkg_v1, pkg_v2).is_ok());
}

fn get_packages(name: &str) -> (Vec<CompiledModule>, Vec<CompiledModule>) {
    let mut path: PathBuf = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    path.push("src/unit_tests/fixtures/upgrade_errors/");
    path.push(format!("{}_v1", name));

    let pkg_v1 = BuildConfig::new_for_testing()
        .build(&path)
        .unwrap()
        .into_modules();

    let mut path: PathBuf = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    path.push("src/unit_tests/fixtures/upgrade_errors/");
    path.push(format!("{}_v2", name));

    let pkg_v2 = BuildConfig::new_for_testing()
        .build(&path)
        .unwrap()
        .into_modules();

    (pkg_v1, pkg_v2)
}
