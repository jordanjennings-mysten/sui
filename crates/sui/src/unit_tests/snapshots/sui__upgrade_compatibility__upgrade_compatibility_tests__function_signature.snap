---
source: crates/sui/src/unit_tests/upgrade_compatibility_tests.rs
expression: err.to_string()
---
error: Function signature mismatch
  ┌─ /Users/jordanjennings/code/sui/crates/sui/src/unit_tests/fixtures/upgrade_errors/function_signature_v2/sources/UpgradeErrors.move:9:38
  │
9 │     public fun func_with_wrong_param(a: u32): u64 {
  │                                      ^ Function 'func_with_wrong_param' unexpected parameter u32 at position 0, expected u64
  │
  = Functions are part of a module's public interface and cannot be changed during an upgrade, restore the original function's parameters for function 'func_with_wrong_param'.

error: Function signature mismatch
   ┌─ /Users/jordanjennings/code/sui/crates/sui/src/unit_tests/fixtures/upgrade_errors/function_signature_v2/sources/UpgradeErrors.move:13:42
   │
13 │     public fun func_with_wrong_return(): u32 {
   │                                          ^^^ Module 'upgrades' function 'func_with_wrong_return' has an unexpected return type u32, expected u64
   │
   = Functions are part of a module's public interface and cannot be changed during an upgrade, restore the original function's return types for function 'func_with_wrong_return'.

error: Function signature mismatch
   ┌─ /Users/jordanjennings/code/sui/crates/sui/src/unit_tests/fixtures/upgrade_errors/function_signature_v2/sources/UpgradeErrors.move:17:49
   │
17 │     public fun func_with_wrong_param_and_return(a: u32): u32 {
   │                                                 ^ Function 'func_with_wrong_param_and_return' unexpected parameter u32 at position 0, expected u64
   │
   = Functions are part of a module's public interface and cannot be changed during an upgrade, restore the original function's parameters for function 'func_with_wrong_param_and_return'.

error: Function signature mismatch
   ┌─ /Users/jordanjennings/code/sui/crates/sui/src/unit_tests/fixtures/upgrade_errors/function_signature_v2/sources/UpgradeErrors.move:17:58
   │
17 │     public fun func_with_wrong_param_and_return(a: u32): u32 {
   │                                                          ^^^ Module 'upgrades' function 'func_with_wrong_param_and_return' has an unexpected return type u32, expected u64
   │
   = Functions are part of a module's public interface and cannot be changed during an upgrade, restore the original function's return types for function 'func_with_wrong_param_and_return'.

error: Function signature mismatch
   ┌─ /Users/jordanjennings/code/sui/crates/sui/src/unit_tests/fixtures/upgrade_errors/function_signature_v2/sources/UpgradeErrors.move:21:16
   │
21 │     public fun func_with_wrong_param_length(a: u64): u64 {
   │                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Function 'func_with_wrong_param_length' expected 2 parameters, have 1
   │
   = Functions are part of a module's public interface and cannot be changed during an upgrade, restore the original function's parameters for function 'func_with_wrong_param_length', expected 2 parameters.

error: Function signature mismatch
   ┌─ /Users/jordanjennings/code/sui/crates/sui/src/unit_tests/fixtures/upgrade_errors/function_signature_v2/sources/UpgradeErrors.move:25:16
   │
25 │     public fun func_with_wrong_return_length(): u64 {
   │                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Function 'func_with_wrong_return_length' expected to have 2 return type(s), have 1
   │
   = Functions are part of a module's public interface and cannot be changed during an upgrade, restore the original function's return types for function 'func_with_wrong_return_length'.


Upgrade failed, this package requires changes to be compatible with the existing package. It's upgrade policy is set to 'Compatible'.
